pkg/controllers/deprovisioning/controller.go

```diff
@@ -297,26 +301,40 @@ func (c *Controller) waitForDeletion(ctx context.Context, nodeClaim *v1beta1.Nod
 	}
 }
 
-func (c *Controller) setNodesUnschedulable(ctx context.Context, isUnschedulable bool, candidates ...*Candidate) error {
+// requireDisruptionNoScheduleTaint will either add/remove the karpenter.sh/disruption taint from the candidates.
+// This is used to enforce no taints at the beginning of deprovisioning, and
+// to add/remove taints while executing a deprovisioning action.
+func (c *Controller) requireDisruptionNoScheduleTaint(ctx context.Context, addTaint bool, nodes ...*state.StateNode) error {
 	var multiErr error
-	for _, cn := range candidates {
+	for _, n := range nodes {
 		node := &v1.Node{}
-		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: cn.Node.Name}, node); err != nil {
+		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: n.Name()}, node); client.IgnoreNotFound(err) != nil {
 			multiErr = multierr.Append(multiErr, fmt.Errorf("getting node, %w", err))
 		}
 
-		// node is being deleted already, so no need to un-cordon
-		if !isUnschedulable && !node.DeletionTimestamp.IsZero() {
+		// If the node already has the taint, continue to the next
+		_, ok := lo.Find(node.Spec.Taints, func(taint v1.Taint) bool {
+			return v1beta1.DisruptionNoScheduleTaint.MatchTaint(&taint)
```

Comment:
Let's double-check that MatchTaint has the correct logic for handling, key, value, effect. IIRC, it might not be correct